flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriverTest.java

```diff
@@ -0,0 +1,295 @@
+				});
+
+				// send onAdded event
+				getPodCallbackHandler().onAdded(Collections.singletonList(pod));
+
+				// make sure finishing validation
+				validationFuture.get(TIMEOUT_SEC, TimeUnit.SECONDS);
+			});
+		}};
+	}
+
+	@Test
+	public void testOnPodModified() throws Exception {
+		new Context() {{
+			testOnPodTerminated((pod) -> getPodCallbackHandler().onModified(pod));
+		}};
+	}
+
+	@Test
+	public void testOnPodDeleted() throws Exception {
+		new Context() {{
+			testOnPodTerminated((pod) -> getPodCallbackHandler().onDeleted(pod));
+		}};
+	}
+
+	@Test
+	public void testOnError() throws Exception {
+		new Context() {{
+			testOnPodTerminated((pod) -> getPodCallbackHandler().onError(pod));
+		}};
+	}
+
+	@Test
+	public void testFatalHandleError() throws Exception {
+		new Context() {{
+			final CompletableFuture<Throwable> onErrorFuture = new CompletableFuture<>();
+			resourceEventHandlerBuilder.setOnErrorConsumer(onErrorFuture::complete);
+
+			runTest(() -> {
+				final Throwable testingError = new Throwable("testing error");
+				getPodCallbackHandler().handleFatalError(testingError);
+				assertThat(onErrorFuture.get(TIMEOUT_SEC, TimeUnit.SECONDS), is(testingError));
+			});
+		}};
+	}
+
+	@Test
+	public void testPodCreationInterval() throws Exception {
+		new Context() {{
+			final AtomicInteger createPodCount = new AtomicInteger(0);
+			final List<CompletableFuture<Long>> createPodTimeFutures = new ArrayList<>();
+			createPodTimeFutures.add(new CompletableFuture<>());
+			createPodTimeFutures.add(new CompletableFuture<>());
+
+			flinkKubeClientBuilder.setCreateTaskManagerPodFunction((ignore) -> {
+				int idx = createPodCount.getAndIncrement();
+				if (idx < createPodTimeFutures.size()) {
+					createPodTimeFutures.get(idx).complete(System.currentTimeMillis());
+				}
+				return FutureUtils.completedExceptionally(new Throwable("testing error"));
+			});
+
+			runTest(() -> {
+				// re-request resource on pod creation failed
+				runInMainThread(() -> getDriver().requestResource(TASK_EXECUTOR_PROCESS_SPEC)
+						.whenComplete((ignore1, ignore2) -> getDriver().requestResource(TASK_EXECUTOR_PROCESS_SPEC)));
+
+				// validate trying creating pod twice, with proper interval
+				long t1 = createPodTimeFutures.get(0).get(TIMEOUT_SEC, TimeUnit.SECONDS);
+				long t2 = createPodTimeFutures.get(1).get(TIMEOUT_SEC, TimeUnit.SECONDS);
+				assertThat((t2 - t1), greaterThanOrEqualTo(POD_CREATION_INTERVAL.toMilliseconds()));
+			});
+		}};
+	}
+
+	@Override
+	protected ResourceManagerDriverTestBase<KubernetesWorkerNode>.Context createContext() {
+		return new Context();
+	}
+
+	private class Context extends ResourceManagerDriverTestBase<KubernetesWorkerNode>.Context {
+		private final KubernetesPod previousAttemptPod = new TestingKubernetesPod(CLUSTER_ID + "-taskmanager-1-1");
+
+		final CompletableFuture<PodCallbackHandler> setWatchPodsAndDoCallbackFuture = new CompletableFuture<>();
```

Comment:
Well, I was not intended to prevent these variables from being accessed by the test cases.
But I think you are right, it does not hurt to make the `private`. They are not accessed from anywhere else, and one could change that once such accesses are needed.